### ISSUE TITLE
lsp: unable to import types in JSDoc

### ISSUE THREAD

**Issue by jaydenseric** [denoland/deno#11362]
In regular TypeScript projects, within a `.js` file with type checking enabled via `// @ts-check` you can import types from other files within JSDoc comments. For example…

In `a.js`:

```js
// @ts-check

/**
 * Description for A.
 * @param {import("./b.js").default} b The B.
 */
export default function a(b) {
  return b;
}
```

In `b.js`:

```js
// @ts-check

/**
 * Description for B.
 * @type {boolean}
 */
export default true;
```

Note that the parameter type is able to be imported and checked:

<img width="453" alt="Screen Shot 2021-07-12 at 10 43 26 am" src="https://user-images.githubusercontent.com/1754873/125215375-b1f31300-e2fe-11eb-868d-579603538688.png">

With exactly the same setup in a Deno project, the parameter type cannot be imported or checked:

<img width="447" alt="Screen Shot 2021-07-12 at 10 43 07 am" src="https://user-images.githubusercontent.com/1754873/125215443-db13a380-e2fe-11eb-9705-318b642ce12b.png">

This not working in Deno, with no open issues about it has really made the past few days trying to get an isomorphic codebase with a mixture of `.ts`, `.tsx`, and `.js` files a nightmare, because it's hard to tell if it's you or Deno that's doing the wrong thing. I _think_ if this Deno bug were fixed I'd finally have a workable solution!

**Comment by jaydenseric** [comment]
Has any investigation on this happened yet; is this an easy or hard bug to fix?

If it's a low-hanging fruit, patching this would be a game changer for those of us trying to develop and publish isomorphic library code with type safety. Type checking is practically impossible for universal browser/Deno modules until this is fixed.

**Comment by jaydenseric** [comment]
Here is an example of a type importable in a `.ts` file:

```ts
import type { Context } from "https://deno.land/x/oak@v9.0.1/mod.ts";
```

<img width="642" alt="Screen Shot 2021-10-04 at 6 27 13 pm" src="https://user-images.githubusercontent.com/1754873/135811110-c0189488-0208-465f-a75c-53678da42d8b.png">

As you can see, attempting to import it for use in a JSDoc `@typedef` doesn't work:

```js
// @ts-check
/** @typedef {import("https://deno.land/x/oak@v9.0.1/mod.ts").Context} Context */
```

<img width="740" alt="Screen Shot 2021-10-04 at 6 29 05 pm" src="https://user-images.githubusercontent.com/1754873/135811202-85d430fe-66a3-4154-baa3-42c1bc702f7a.png">

It is a really mysterious thing why importing types in JSDoc sometimes works, and sometimes doesn't. It seems to never work when you import a type from a `.ts` file into JSDoc in a `.js` file though.

This issue hasn't been investigated or fixed in the nearly 3 months since it was filed in June 😢

I'm sure in time Deno will shape up to be the dream JS/TS runtime and I appreciate the efforts of the Deno collaborators and contributors, but IMO the Deno team needs to focus more on unblocking package authors that are working hard pro-bono to build out packages to grow the Deno ecosystem. Bugs that are showstoppers for type checking (e.g. https://github.com/denoland/deno/issues/11362 , https://github.com/denoland/deno/issues/12143 , https://github.com/denoland/deno/issues/11786#issuecomment-921491332), code coverage (e.g. https://github.com/denoland/deno/issues/11327), etc. need to be dealt with with a sense of urgency instead of working on enhancements or nice-to-haves like Node.js ecosystem compatibility. That way both the Deno runtime and a useful ecosystem of packages can be developed in parallel, and the Deno API can evolve sooner to meet the needs of package authors.

**Comment by kitsonk** [comment]
Even though #13156 came long after, @jespertheend was able to isolate it in that issue, and changes are being made to fix the issue, so we will use #13156 for tracking purposes.

Closing this as a duplicate.
